### PR TITLE
Add the section_for_each and set_readonly functions for ini file (for CoC Compatibility)

### PR DESCRIPTION
**Pull request by ChugunovRoman**

### CODE REVIEW

File: src/xrCore/xr_ini.h [hunk 1]
```diff
@@ -22,6 +22,13 @@ class IReader;
 class XRCORE_API CInifile
 {
 public:
+    enum
+    {
+        eSaveAtEnd = 1 << 0,
+        eReadOnly = 1 << 1,
+        eOverrideNames = 1 << 2,
+    };
+    Flags8 m_flags;
```
Comment (Xottab-DUTY) [inline src/xrCore/xr_ini.h]: Don't make it public. It's ok to make enum public, but not the variable.

File: src/xrServerEntities/script_ini_file.cpp [hunk 2]
```diff
@@ -213,6 +217,22 @@ void CScriptIniFile::set_override_names(bool b)
     inherited::set_override_names(b);
 }
 
+void CScriptIniFile::section_for_each(const luabind::functor<void>& functor)
+{
+    using sections_type = CInifile::Root;
+    sections_type& sections = this->sections();
```
Comment (Xottab-DUTY) [inline src/xrServerEntities/script_ini_file.cpp]: Never use this, only if it's really necessary..
```suggestion
    sections_type& sections = sections();
```

Comment (Xottab-DUTY) [reply src/xrServerEntities/script_ini_file.cpp]: Also, is it possible to make it `const`?

File: src/xrServerEntities/script_ini_file.h [hunk 3]
```diff
@@ -8,7 +8,12 @@
 
 #pragma once
 
+#include "pch_script.h"
 #include "script_token_list.h"
+#include "xrScriptEngine/Functor.hpp"
+
+using namespace luabind;
+using namespace luabind::policy;
```
Comment (Xottab-DUTY) [inline src/xrServerEntities/script_ini_file.h]: Never use `using namespace` in the global headers.

File: src/xrServerEntities/script_ini_file_script.cpp [hunk 4]
```diff
@@ -101,7 +102,10 @@ static void CScriptIniFile_Export(lua_State* luaState)
             .def("remove_line", &CScriptIniFile::remove_line)
             .def("set_override_names", &CScriptIniFile::set_override_names)
             .def("section_count", &CScriptIniFile::section_count)
+            .def("section_for_each", &CScriptIniFile::section_for_each)
+            .def("set_readonly", &CScriptIniFile::set_readonly)
             //Alundaio: END
+            .def("fname", REMOVE_NOEXCEPT(&CScriptIniFile::fname))
```
Comment (Xottab-DUTY) [inline src/xrServerEntities/script_ini_file_script.cpp]: Looks like wrong thing to do and we should remove `noexcept` specifier *literally*. (comment it out in `fname` function)

File: src/xrServerEntities/script_ini_file.h [hunk 5]
```diff
@@ -8,7 +8,12 @@
 
 #pragma once
 
+#include "pch_script.h"
```
Comment (Xottab-DUTY) [inline src/xrServerEntities/script_ini_file.h]: NEVER do this. This file should only be included into .cpp's which need this file.

**Review by Xottab-DUTY — CHANGES_REQUESTED** [review]

File: src/xrCore/xr_ini.h (hunk 1 above)
Comment (ChugunovRoman) [reply src/xrCore/xr_ini.h]: Fixed

File: src/xrServerEntities/script_ini_file.h (hunk 5 above)
Comment (ChugunovRoman) [reply src/xrServerEntities/script_ini_file.h]: Okay, I removed. But for what this header is needed?

File: src/xrServerEntities/script_ini_file_script.cpp (hunk 4 above)
Comment (ChugunovRoman) [reply src/xrServerEntities/script_ini_file_script.cpp]: I just removed `REMOVE_NOEXCEPT`.

File: src/xrServerEntities/script_ini_file.h (hunk 3 above)
Comment (ChugunovRoman) [reply src/xrServerEntities/script_ini_file.h]: Fixed

File: src/xrServerEntities/script_ini_file.cpp (hunk 2 above)
Comment (ChugunovRoman) [reply src/xrServerEntities/script_ini_file.cpp]: Fixed

File: src/xrServerEntities/script_ini_file_script.cpp (hunk 4 above)
Comment (ChugunovRoman) [reply src/xrServerEntities/script_ini_file_script.cpp]: The `fname` function has `noexcept` specifier, but it's not prevent build fail

File: src/xrServerEntities/script_ini_file.h (hunk 5 above)
Comment (Xottab-DUTY) [reply src/xrServerEntities/script_ini_file.h]: pch_script.h is a precompiled header

File: src/xrServerEntities/script_ini_file_script.cpp (hunk 4 above)
Comment (Xottab-DUTY) [reply src/xrServerEntities/script_ini_file_script.cpp]: Yes, the compilation fails exactly because of that. `noexcept` should be removed from the function itself. (commented out)

File: src/xrGame/xrGame.vcxproj [hunk 6]
```diff
@@ -3530,7 +3530,7 @@
     <ProjectReference Include="$(xrExternals)luabind.vcxproj">
       <Project>{ccd4afae-aa10-42c6-a452-fdee497ccdf1}</Project>
     </ProjectReference>
-    <ProjectReference Include="..\xrUICore\xrUICore.vcxproj">
+    <ProjectReference Include="$(SolutionDir)xrUICore\xrUICore.vcxproj">
```
Comment (Xottab-DUTY) [inline src/xrGame/xrGame.vcxproj]: This change is not related to this PR, please, remove

File: src/xrScriptEngine/xrScriptEngine.vcxproj [hunk 7]
```diff
@@ -84,7 +84,7 @@
     <ProjectReference Include="$(xrExternals)luabind.vcxproj">
       <Project>{ccd4afae-aa10-42c6-a452-fdee497ccdf1}</Project>
     </ProjectReference>
-    <ProjectReference Include="..\..\Externals\LuaJIT-proj\LuaJIT.vcxproj">
+    <ProjectReference Include="$(xrExternals)\LuaJIT-proj\LuaJIT.vcxproj">
```
Comment (Xottab-DUTY) [inline src/xrScriptEngine/xrScriptEngine.vcxproj]: This change is not related to this PR, please, remove

File: src/xrServerEntities/script_ini_file_script.cpp [hunk 8]
```diff
@@ -9,6 +9,7 @@
 #include "pch_script.h"
 #include "script_ini_file.h"
 #include "xrScriptEngine/ScriptExporter.hpp"
+#include "Common/object_type_traits.h"
```
Comment (Xottab-DUTY) [inline src/xrServerEntities/script_ini_file_script.cpp]: Not needed anymore

File: src/xrServerEntities/script_ini_file.h [hunk 9]
```diff
@@ -54,6 +55,8 @@ class CScriptIniFile : public CInifile
     void save_at_end(bool b);
     void remove_line(pcstr S, pcstr L);
     void set_override_names(bool b);
+    void section_for_each(const luabind::functor<void>& functor) const;
+    void set_readonly(bool b);
```
Comment (Xottab-DUTY) [inline src/xrServerEntities/script_ini_file.h]: This is not needed now, inherited will be used automatically
```suggestion
```

File: src/xrServerEntities/script_ini_file.h [hunk 10]
```diff
@@ -9,6 +9,7 @@
 #pragma once
 
 #include "script_token_list.h"
+#include "xrScriptEngine/Functor.hpp"
```
Comment (Xottab-DUTY) [inline src/xrServerEntities/script_ini_file.h]: Using this header looks bad, maybe it may affect compilation speed...

We can just use lambda script_ini_file_script.cpp :)

File: src/xrServerEntities/script_ini_file_script.cpp [hunk 11]
```diff
@@ -101,7 +102,10 @@ static void CScriptIniFile_Export(lua_State* luaState)
             .def("remove_line", &CScriptIniFile::remove_line)
             .def("set_override_names", &CScriptIniFile::set_override_names)
             .def("section_count", &CScriptIniFile::section_count)
+            .def("section_for_each", &CScriptIniFile::section_for_each)
```
Comment (Xottab-DUTY) [inline src/xrServerEntities/script_ini_file_script.cpp]: Let's use lambda right here :)

```suggestion
            .def("section_for_each", +[](CScriptIniFile* self, const luabind::functor<void>& functor)
            {
                using sections_type = CInifile::Root;
                const sections_type& sections = self->sections();

                for (auto& section : sections)
                {
                    functor(section->Name.c_str());
                }
            })
```

File: src/xrCore/xr_ini.h [hunk 12]
```diff
@@ -83,7 +83,7 @@ class XRCORE_API CInifile
     void save_as(IWriter& writer, bool bcheck = false) const;
     void set_override_names(bool b) noexcept { m_flags.set(eOverrideNames, b); }
     void save_at_end(bool b) noexcept { m_flags.set(eSaveAtEnd, b); }
-    pcstr fname() const noexcept { return m_file_name; };
+    pcstr fname() const { return m_file_name; };
```
Comment (Xottab-DUTY) [inline src/xrCore/xr_ini.h]: Ну я же попросил закомментировать, а не удалять...
```suggestion
    pcstr fname() const /*noexcept*/ { return m_file_name; };
```

File: src/xrServerEntities/script_ini_file.cpp [hunk 13]
```diff
@@ -6,13 +6,17 @@
...
+#include "pch_script.h"
 #include "StdAfx.h"
 #include "script_ini_file.h"
 #include "xrScriptEngine/script_engine.hpp"
 #include "ai_space.h"
 #include "object_factory.h"
 #include "xrCore/xrDebug.h"
 
+using namespace luabind;
+using namespace luabind::policy;
```
Comment (Xottab-DUTY) [inline src/xrServerEntities/script_ini_file.cpp]: Since, everything was moved inside `script_ini_file_script.cpp`, please, revert this file 👍

File: src/utils/xrSE_Factory/xrSE_Factory.vcxproj [hunk 14]
```diff
@@ -173,7 +173,10 @@
       <PrecompiledHeaderFile>pch_script.h</PrecompiledHeaderFile>
       <PrecompiledHeaderOutputFile>$(IntDir)$(ProjectName)_script.pch</PrecompiledHeaderOutputFile>
     </ClCompile>
-    <ClCompile Include="..\..\xrServerEntities\script_ini_file.cpp" />
+    <ClCompile Include="..\..\xrServerEntities\script_ini_file.cpp" >
+      <PrecompiledHeaderFile>pch_script.h</PrecompiledHeaderFile>
+      <PrecompiledHeaderOutputFile>$(IntDir)$(ProjectName)_script.pch</PrecompiledHeaderOutputFile>
+    </ClCompile>
```
Comment (Xottab-DUTY) [inline src/utils/xrSE_Factory/xrSE_Factory.vcxproj]: This file can be reverted too

File: src/xrServerEntities/script_ini_file_script.cpp [hunk 15]
```diff
@@ -101,7 +102,19 @@ static void CScriptIniFile_Export(lua_State* luaState)
             .def("remove_line", &CScriptIniFile::remove_line)
             .def("set_override_names", &CScriptIniFile::set_override_names)
             .def("section_count", &CScriptIniFile::section_count)
+            .def("section_for_each", +[](CScriptIniFile* self, const luabind::functor<void>& functor)
+            {
+                using sections_type = CInifile::Root;
+                sections_type const& sections = self->sections();
```
Comment (Xottab-DUTY) [inline src/xrServerEntities/script_ini_file_script.cpp]: We use left const.
```suggestion
                const sections_type& sections = self->sections();
```

**Review by Xottab-DUTY — APPROVED** [review]

**Comment by Xottab-DUTY** [comment]
Awesome, thanks!

**Review by Xottab-DUTY — APPROVED** [review]
